### PR TITLE
ARROW-1280: [C++] add fixed size list type

### PR DESCRIPTION
**Pull request by bkietz**
- FixedSizeListType, parameterized by the list size (similar to FixedSizeBinaryType), constructed with `fixed_size_list(<value field or type>, list_size)`
- FixedSizeListArray, including Equal and pretty print impls
- FixedSizeListBuilder, works with MakeBuilder

### CODE REVIEW

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -522,6 +522,44 @@ class ARROW_EXPORT ListArray : public Array {
...
+  std::shared_ptr<Array> values() const;
+
+  std::shared_ptr<DataType> value_type() const;
+
+  // Neither of these functions will perform boundschecking
+  int32_t value_offset(int64_t i) const {
+    i += data_->offset;
+    return static_cast<int32_t>(list_size_ * i);
+  }
+  int32_t value_length(...) const { return list_size_; }
```
Comment (fsaintjacques) [inline cpp/src/arrow/array.h]: I think I prefer the explicit `int64_t i` as parameter. A bit better for the ABI.

File: cpp/src/arrow/array.h [hunk 2]
```diff
@@ -522,6 +522,44 @@ class ARROW_EXPORT ListArray : public Array {
   std::shared_ptr<Array> values_;
 };
 
+// ----------------------------------------------------------------------
+// FixedSizeListArray
+
+/// Concrete Array class for list data
+class ARROW_EXPORT FixedSizeListArray : public Array {
```
Comment (fsaintjacques) [inline cpp/src/arrow/array.h]: is it worth having common abstract class?

**Review by fsaintjacques — COMMENTED** [review]
It think it's worth adding support in the integration tests as Java already support this type:

https://github.com/apache/arrow/blob/master/integration/integration_test.py

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/array.h]: Even though it's independent of index?

File: cpp/src/arrow/array.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/array.h]: There's not much opportunity for code reuse in arrays/builders. The only advantage I can think of would be easier reuse of code which consumes `ListArray` (but only code which doesn't directly access the offsets buffer). Those algorithms could use `ArrayData BasicListArray::GetView(int64_t i)` or similar. This seems like speculative generality to me; I'd be more open to doing it if you can think of existing code which would benefit from being refactored this way.

`FixedSizeListScalar` and `ListScalar` are identical but also trivial.

File: cpp/src/arrow/array/builder_nested.h [hunk 3]
```diff
@@ -87,6 +87,72 @@ class ARROW_EXPORT ListBuilder : public ArrayBuilder {
...
+  Status Append();
+
+  /// \brief Vector append
+  ///
+  /// If passed, valid_bytes wil be read and any zero byte
+  /// will cause the corresponding slot to be null
+  ///
+  /// This function affects only the validity bitmap; the child values must be appended
+  /// using the child array builder. This includes appending nulls for null lists.
+  /// XXX this restriction is confusing, should this method be omitted?
```
Comment (pitrou) [inline cpp/src/arrow/array/builder_nested.h]: This is the same kind of restriction as for other nested types, right?

File: cpp/src/arrow/array/builder_nested.cc [hunk 4]
```diff
@@ -142,6 +142,80 @@ ArrayBuilder* ListBuilder::value_builder() const {
...
+    int32_t list_size)
+    : ArrayBuilder(fixed_size_list(value_builder->type(), list_size), pool),
+      list_size_(list_size),
+      value_builder_(value_builder) {}
+
+FixedSizeListBuilder::FixedSizeListBuilder(
+    MemoryPool* pool, std::shared_ptr<ArrayBuilder> const& value_builder,
+    const std::shared_ptr<DataType>& type)
+    : ArrayBuilder(type, pool),
+      list_size_(static_cast<const FixedSizeListType*>(type.get())->list_size()),
```
Comment (pitrou) [inline cpp/src/arrow/array/builder_nested.cc]: Should use `checked_cast` here.

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/array.h]: I would rather remove the argument. Or you can provide two separate overrides. A variadic function looks bizarre here.

File: cpp/src/arrow/array.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/array.h]: Agreed with @bkietz. Also code reuse can be achieved through templating (we generally care about performing non-virtual calls anyway).

File: cpp/src/arrow/array.h [hunk 5]
```diff
@@ -522,6 +522,44 @@ class ARROW_EXPORT ListArray : public Array {
   std::shared_ptr<Array> values_;
 };
 
+// ----------------------------------------------------------------------
+// FixedSizeListArray
+
+/// Concrete Array class for list data
```
Comment (pitrou) [inline cpp/src/arrow/array.h]: "fixed-size list data", probably.

File: cpp/src/arrow/array.cc [hunk 6]
```diff
@@ -839,6 +878,22 @@ struct ValidateVisitor {
...
   }
 
+  Status Visit(const FixedSizeListArray& array) {
+    if (array.length() < 0) {
+      return Status::Invalid("Length was negative");
+    }
+    if (!array.values()) {
+      return Status::Invalid("values was null");
+    }
+    if (array.values()->length() != array.length() * array.value_length()) {
```
Comment (pitrou) [inline cpp/src/arrow/array.cc]: +1

File: cpp/src/arrow/array.h [hunk 7]
```diff
@@ -522,6 +522,44 @@ class ARROW_EXPORT ListArray : public Array {
...
+class ARROW_EXPORT FixedSizeListArray : public Array {
+ public:
+  using TypeClass = FixedSizeListType;
+
+  explicit FixedSizeListArray(const std::shared_ptr<ArrayData>& data);
+
+  FixedSizeListArray(const std::shared_ptr<DataType>& type, int64_t length,
+                     const std::shared_ptr<Buffer>& value_offsets,
```
Comment (pitrou) [inline cpp/src/arrow/array.h]: Why is this constructor taking a value offsets buffer?

File: cpp/src/arrow/array.cc [hunk 8]
```diff
@@ -295,6 +295,45 @@ std::shared_ptr<DataType> ListArray::value_type() const {
...
+  SetData(data);
+}
+
+FixedSizeListArray::FixedSizeListArray(const std::shared_ptr<DataType>& type,
+                                       int64_t length,
+                                       const std::shared_ptr<Buffer>& value_offsets,
+                                       const std::shared_ptr<Array>& values,
+                                       const std::shared_ptr<Buffer>& null_bitmap,
+                                       int64_t null_count, int64_t offset) {
+  auto internal_data =
```
Comment (pitrou) [inline cpp/src/arrow/array.cc]: Perhaps DCHECK that `type->value_type()` matches `values->type()`?

File: cpp/src/arrow/array-list-test.cc [hunk 9]
```diff
@@ -332,4 +332,219 @@ TEST_F(TestListArray, TestBuilderPreserveFieleName) {
...
+  ValidateBasicFixedSizeListArray(result_.get(), values, is_valid);
+}
+
+TEST_F(TestFixedSizeListArray, BulkAppend) {
+  std::vector<int32_t> values = {0, 1, 2, 3, 4, 5};
+  std::vector<uint8_t> is_valid = {1, 0, 1};
+
+  Int32Builder* vb = checked_cast<Int32Builder*>(builder_->value_builder());
+
+  ASSERT_OK(builder_->Reserve(values.size() / list_size()));
```
Comment (pitrou) [inline cpp/src/arrow/array-list-test.cc]: Is it required to `Reserve` explicitly here? The builder should reserve automatically when the safe append methods are called.

File: cpp/src/arrow/array-list-test.cc [hunk 10]
```diff
@@ -332,4 +332,219 @@ TEST_F(TestListArray, TestBuilderPreserveFieleName) {
...
+  auto list_type_with_name = fixed_size_list(field("counts", int32()), list_size());
+
+  std::unique_ptr<ArrayBuilder> tmp;
+  ASSERT_OK(MakeBuilder(pool_, list_type_with_name, &tmp));
+  builder_.reset(checked_cast<FixedSizeListBuilder*>(tmp.release()));
+
+  ASSERT_OK(builder_->AppendValues(4));
+
+  std::shared_ptr<Array> list_array;
+  ASSERT_OK(builder_->Finish(&list_array));
```
Comment (pitrou) [inline cpp/src/arrow/array-list-test.cc]: IIUC, this builds an invalid FixedSizeListArray? No values are appended to the child array.

**Review by pitrou — COMMENTED** [review]
Thank you. Overall this looks good. Still some comments below.

Side question: should support for `FixedSizeListType` added to `ArrayFromJSON`? Perhaps as a separate PR?

File: cpp/src/arrow/array/builder_nested.h (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/array/builder_nested.h]: It seemed more confusing to me compared to List, which doesn't require insertion into the child builder in the case of nulls. By contrast, FixedSizeList must have `N` insertions even for null entries

Comment (pitrou) [reply cpp/src/arrow/array/builder_nested.h]: For the record, that's required on Struct arrays too.

File: cpp/src/arrow/array.h (hunk 7 above)
Comment (bkietz) [reply cpp/src/arrow/array.h]: Good catch

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/array.h]: Allowing it to be called with an argument means I don't have to rewrite pretty-print :)

File: cpp/src/arrow/array/builder_nested.h [hunk 11]
```diff
@@ -87,6 +87,72 @@ class ARROW_EXPORT ListBuilder : public ArrayBuilder {
   Status AppendNextOffset(int64_t num_repeats);
 };
 
+// ----------------------------------------------------------------------
+// FixedSizeList builder
+
+/// \class FixedSizeListBuilder
+/// \brief Builder class for fixed-length list array value types
+///
+/// A note on types.  Per arrow/type.h all types in the c++ implementation are
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/array/builder_nested.h]: I'll just delete it

File: cpp/src/arrow/array-list-test.cc (hunk 10 above)
Comment (bkietz) [reply cpp/src/arrow/array-list-test.cc]: It does build an invalid array, but the point of this test is just to check that the field name is correctly propagated to the built array

File: cpp/src/arrow/array-list-test.cc (hunk 9 above)
Comment (bkietz) [reply cpp/src/arrow/array-list-test.cc]: You're correct. It's just copypasta from the List test cases; I'll remove it.

**Comment by bkietz** [comment]
@pitrou Added [ArrayFromJSON impl](https://github.com/apache/arrow/pull/4278/files#diff-015ed4b6849ed6e64e25bba42aa1d29eR623)

File: cpp/src/arrow/ipc/json-internal.cc [hunk 12]
```diff
@@ -556,6 +556,12 @@ class ArrayWriter {
     return WriteChildren(type.children(), {array.values()});
   }
 
+  Status Visit(const FixedSizeListArray& array) {
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/ipc/json-internal.cc]: How's this: https://github.com/apache/arrow/pull/4278/files#diff-4e2d1d19217cc5481a51f3aa0b65b38eR193

File: cpp/src/arrow/compare.cc [hunk 13]
```diff
@@ -576,12 +584,22 @@ class ArrayEqualsVisitor : public RangeEqualsVisitor {
       return Status::OK();
     }
 
+    // XXX this looks like an error where the author expected RangeEquals to take
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/compare.cc]: Confirmed, [this is a bug](https://issues.apache.org/jira/browse/ARROW-5299).

File: cpp/src/arrow/array.h (hunk 2 above)
Comment (fsaintjacques) [reply cpp/src/arrow/array.h]: :+1:

**Comment by pitrou** [comment]
@kou There is a meson-related error on Travis-CI, could you advise?

**Comment by pitrou** [comment]
Looks like there's also a MinGW test failure:
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/24461942/job/mper2jipn6kj5bxj

```
Error: test: csv.gz(TableTest::#save and .load::path:::format::load: auto detect): Arrow::Error::Io: [csv-reader][read]: IOError: zlib inflate failed: invalid distance too far back
c:/Ruby26/lib/ruby/gems/2.6.0/gems/gobject-introspection-3.3.6/lib/gobject-introspection/loader.rb:616:in `invoke'
c:/Ruby26/lib/ruby/gems/2.6.0/gems/gobject-introspection-3.3.6/lib/gobject-introspection/loader.rb:616:in `invoke'
c:/Ruby26/lib/ruby/gems/2.6.0/gems/gobject-introspection-3.3.6/lib/gobject-introspection/loader.rb:533:in `block in define_method'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:158:in `block (2 levels) in load_from_path'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:147:in `block (2 levels) in wrap_input'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:140:in `open_encoding_convert_stream'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:146:in `block in wrap_input'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:125:in `block in open_decompress_input'
C:/projects/arrow/ruby/red-arrow/lib/arrow/block-closable.rb:25:in `open'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:124:in `open_decompress_input'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:145:in `wrap_input'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:157:in `block in load_from_path'
C:/projects/arrow/ruby/red-arrow/lib/arrow/block-closable.rb:25:in `open'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:156:in `load_from_path'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:39:in `load'
C:/projects/arrow/ruby/red-arrow/lib/arrow/csv-loader.rb:26:in `load'
C:/projects/arrow/ruby/red-arrow/lib/arrow/table-loader.rb:158:in `load_as_csv'
C:/projects/arrow/ruby/red-arrow/lib/arrow/table-loader.rb:50:in `load'
C:/projects/arrow/ruby/red-arrow/lib/arrow/table-loader.rb:22:in `load'
C:/projects/arrow/ruby/red-arrow/lib/arrow/table.rb:27:in `load'
C:/projects/arrow/ruby/red-arrow/test/test-table.rb:503:in `block (5 levels) in <class:TableTest>'
```

**Comment by pitrou** [comment]
Ok, apparently the Travis-CI failure is unrelated:
https://issues.apache.org/jira/browse/ARROW-5306

**Review by pitrou — APPROVED** [review]
LGTM.

**Comment by pitrou** [comment]
Turns out the MinGW failure was sporadic. Merging.

**Comment by pitrou** [comment]
Thanks @bkietz !

@fsaintjacques Integration tests are tracked in https://issues.apache.org/jira/browse/ARROW-1278.

**Comment by kou** [comment]
If the MinGW failure is caused again, I'll look into it.

**Pull request closed without merging**